base/ca/src/org/dogtagpki/ct/CTEngine.java

```diff
@@ -0,0 +1,704 @@
+
+    /**
+     * Process cert info for Certificate Transparency
+     * 
+     * Check to see if certInfo contains Certificate Transparency poison
+     * extension (from profile containig certTransparencyExtDefaultImpl);
+     *
+     * If not, check if global setting is enabled in CS.cfg:
+     *    ca.certTransparency.enable
+     * and add the poision extension;
+     *
+     * if it does then reach out to the CT log servers to obtain
+     * signed certificate timestamp (SCT) for inclusion of the SCT extension
+     * in the cert to be issued.
+     *
+     */
+    public void process(X509CertInfo certi, ICertificateAuthority ctCA, AuthorityID aid, String algname)
+            throws EBaseException {
+
+        String method = "CTEngine.process: ";
+        ICertificateAuthority ca = ctCA.getCA(aid);
+        String errMsg = "";
+
+        try {
+            if (ctConfig == null) {
+                init();
+            }
+
+            CertificateExtensions exts = (CertificateExtensions) certi.get(X509CertInfo.EXTENSIONS);
+            logger.debug(method + " about to check CT poison");
+            Extension ctPoison = null;
+            try {
+                ctPoison = (Extension) exts.get("1.3.6.1.4.1.11129.2.4.3");
+                logger.debug(method + " CT poison extension found");
+            } catch (Exception e) {
+                logger.debug(method + e.getMessage() + "-- continue");
+                logger.debug(method + " CT poison extension not found");
+            }
+
+            /*
+             * see CTmode for CT mode config info
+             */
+            boolean processCT = false;
+            SCTProcessor.CTmode ct_mode = ctConfig.getCTmode();
+            if (ct_mode == SCTProcessor.CTmode.enabled) {
+                logger.debug(method + "ct_mode is enabled");
+                if (ctPoison == null) {
+                    logger.debug(method + " adding poison ext");
+                    CertUtils.addCTv1PoisonExt(certi);
+                    logger.debug(method + " returned from addCTpoisonExt");
+                }
+                processCT = true;
+            } else if (ct_mode == SCTProcessor.CTmode.perProfile) {
+                logger.debug(method + "ct_mode is perProfile");
+                if (ctPoison != null)
+                    processCT = true;
+            } else { // ct_mode == SCTProcessor.CTmode.disabled
+                logger.debug(method + "ct_mode is disabled");
+                if (ctPoison != null) {
+                    exts.delete("1.3.6.1.4.1.11129.2.4.3");
+                    // debug print without poison ext
+                    CertUtils.printExtensions(exts);
+                    certi.delete(X509CertInfo.EXTENSIONS);
+                    certi.set(X509CertInfo.EXTENSIONS, exts);
+                    logger.debug(method + " ctPoison ext deleted");
+                }
+            }
+
+            if (!processCT) {
+                logger.debug(method + " no CT processing needed");
```

Comment:
Technically the `return` wouldn't make a difference since there's no more code executed after this statement, but if we add one we can move the subsequent code out of the `else` clause and reduce the indentation, making it easier to read.